File: jungle-core/src/ecs/component.rs
use std::{fmt::Display, num::NonZeroU64};

use base64::{Engine, engine::general_purpose};

use crate::ecs::storage::ComponentStorage;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ComponentID {
    id: Option<NonZeroU64>,
}

impl Display for ComponentID {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = if let Some(id) = self.id {
            general_purpose::STANDARD.encode(id.get().to_le_bytes())
        } else {
            String::from("None")
        };
        f.write_str(&format!("Component({})", s))
    }
}

impl ComponentID {
    pub fn from_u64(id: u64) -> Self {
        Self {
            id: NonZeroU64::new(id),
        }
    }
}

pub trait Component: Default {
    type Storage: ComponentStorage<Self>;

    fn name(&self) -> &'static str;
}

File: jungle-core/src/ecs/component.cpp
#include "jungle/core/ecs/component.h"

#include <span>

namespace jungle::core::ecs {

ustr ComponentID::debug() const {
    if (!*this) {
        return ustr{"ComponentID(NONE)"};
    }
    ustr result{"ComponentID("};
    auto range = std::span{reinterpret_cast<const u8 *>(&m_id), sizeof(m_id)};
    result.append(util::base64_encoder_view{range});
    result.append(")");
    return result;
}

};  // namespace jungle::core::ecs
